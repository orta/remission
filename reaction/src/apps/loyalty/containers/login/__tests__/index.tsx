import * as React from "react"
import * as TestUtils from "react-dom/test-utils"
import * as renderer from "react-test-renderer"

jest.mock("isomorphic-fetch")
const fetch = require("isomorphic-fetch").default

import Login from "../index"

describe("login", () => {
  const formConfig = {
    baseUrl: "/loyalty",
    url: "/log_in",
    csrfToken: "csrf",
    facebookPath: "/facebook",
    twitterPath: "/twitter",
  }

  afterEach(fetch.mockReset)

  it("renders the snapshot", () => {
    const login = renderer.create(<Login form={formConfig} />)
    expect(login).toMatchSnapshot()
  })

  it("renders the social buttons with the correct hrefs", () => {
    const login = TestUtils.renderIntoDocument(<Login form={formConfig} />) as Login
    const aTags = TestUtils.scryRenderedDOMComponentsWithTag(login, "a") as HTMLAnchorElement[]
    const facebookLink = aTags.find(tag => tag.href === "/facebook")
    expect(facebookLink).toBeTruthy()
    const twitterLink = aTags.find(tag => tag.href === "/twitter")
    expect(twitterLink).toBeTruthy()
  })

  it("displays correct error message", done => {
    fetch.mockImplementation(() =>
      Promise.resolve({
        url: "http://localhost:3000/log_in?error=Invalid%20email%20or%20password",
        status: 404,
      })
    )

    const login = TestUtils.renderIntoDocument(<Login form={formConfig} />) as Login

    TestUtils.Simulate.submit(TestUtils.findRenderedDOMComponentWithTag(login, "form"))

    // Wait till the next event loop tick to assert on the expected DOM state,
    // because `setState` isn’t going to re-render until the next tick either.
    setImmediate(() => {
      const error = TestUtils.findRenderedDOMComponentWithClass(login, "error")
      expect(error.textContent).toBe("Invalid email or password")
      done()
    })
  })

  it("successful redirects after login", () => {
    fetch.mockImplementation(() => Promise.resolve({ status: 200 }))

    const login = <Login form={formConfig} />
    const renderedComponent = TestUtils.renderIntoDocument(login) as Login

    TestUtils.Simulate.submit(TestUtils.findRenderedDOMComponentWithTag(renderedComponent, "form"))
  })

  it("displays an internal error message", done => {
    fetch.mockImplementation(() => Promise.resolve({ status: 500 }))

    const login = TestUtils.renderIntoDocument(<Login form={formConfig} />) as Login

    TestUtils.Simulate.submit(TestUtils.findRenderedDOMComponentWithTag(login, "form"))

    // Wait till the next event loop tick to assert on the expected DOM state,
    // because `setState` isn’t going to re-render until the next tick either.
    setImmediate(() => {
      const error = TestUtils.findRenderedDOMComponentWithClass(login, "error")
      expect(error.textContent).toBe("Internal Error. Please contact support@artsy.net")
      done()
    })
  })

  it("displays a connection error message from catch", done => {
    fetch.mockImplementation(() => Promise.reject(new Error("Oh noes!")))

    const login = TestUtils.renderIntoDocument(<Login form={formConfig} />) as Login

    TestUtils.Simulate.submit(TestUtils.findRenderedDOMComponentWithTag(login, "form"))

    // Wait till the next event loop tick to assert on the expected DOM state,
    // because `setState` isn’t going to re-render until the next tick either.
    setImmediate(() => {
      const error = TestUtils.findRenderedDOMComponentWithClass(login, "error")
      expect(error.textContent).toBe("Internal Error. Please contact support@artsy.net")
      done()
    })
  })
})
