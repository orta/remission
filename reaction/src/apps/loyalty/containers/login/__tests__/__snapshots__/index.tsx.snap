// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login renders the snapshot 1`] = `
<div
  className="file__LoginContainer-s1d01h8e-0 gwuPpF"
>
  <div
    className="file-s1ekx40u-0 iuEcGN"
    style={undefined}
  >
    
  </div>
  <p
    className="file__Text-s4rtu63-0 eEhqjQ"
  >
    Welcome back, please log in 
    <br />
     to your account.
  </p>
  <form
    action="/log_in"
    method="POST"
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      className="file__StyledInput-s1d01h8e-1 qmhxa file__StyledInput-s2fklig-0 ffQmtI"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      value=""
    />
    <div
      className="file__StyledDiv-s2fklig-2 gnqeba"
    >
      <div
        className="border-container"
      />
      <input
        className="file__BorderlessInput-s2fklig-1 fbjzqZ"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <a
        className="file__StyledTextLink-s18afosz-0 hlqDRc"
        href={undefined}
      >
        Forgot?
      </a>
    </div>
    <button
      className="file__InvertedButton-s2qkmvo-0 MNMDa file__StyledButton-muis26-0 fBHbtu"
    >
      <span>
        Log In
      </span>
    </button>
    <div
      className="file__StyledOrContainer-s1d01h8e-2 iqvrnM"
    >
      <div
        className="file__StyledOrText-s1d01h8e-3 ggmqfI"
      >
        or
      </div>
    </div>
    <a
      className="file-y6ly8y-0 bxHoRp file__StyledButton-muis26-0 fBHbtu"
      href="/facebook"
    >
      <div
        className="file-s1ekx40u-0 bDXcCT"
        style={undefined}
      >
        
      </div>
      <span>
        Log in with Facebook
      </span>
    </a>
    <a
      className="file-cy64ue-0 gjiGIK file__StyledButton-muis26-0 fBHbtu"
      href="/twitter"
    >
      <div
        className="file-s1ekx40u-0 bDXcCT"
        style={undefined}
      >
        
      </div>
      <span>
        Log in with Twitter
      </span>
    </a>
  </form>
  <br />
</div>
`;
